### PR TITLE
Final adaptations for JSON export

### PR DESCRIPTION
**Pull request by rd2**

### CODE REVIEW

File: lib/measures/tbd/README.md [hunk 1]
```diff
@@ -20,14 +20,14 @@ ModelMeasure
...
 **Type:** Boolean,
 **Units:** ,
 **Required:** true,
 **Model Dependent:** false
 
-### Default thermal bridge option to use if not reading TDB.json
+### Default thermal bridge option to use if not reading tbd.json
 e.g. poor, regular, efficient, code
 **Name:** option,
 **Type:** Choice,
```
Comment (rd2) [inline lib/measures/tbd/README.md:33]: Minor cleanup - keeping the lower case version.

File: lib/measures/tbd/measure.rb [hunk 2]
```diff
@@ -39,15 +39,15 @@ def arguments(model)
...
 
     choices = OpenStudio::StringVector.new
     psi = PSI.new
     psi.set.keys.each do |k| choices << k.to_s; end
     option = OpenStudio::Measure::OSArgument.makeChoiceArgument("option", choices, true)
-    option.setDisplayName("Default thermal bridge option to use if not reading TDB.json")
+    option.setDisplayName("Default thermal bridge option to use if not reading tbd.json")
     option.setDescription("e.g. poor, regular, efficient, code")
     option.setDefaultValue("poor (BC Hydro)")
     args << option
```
Comment (rd2) [inline lib/measures/tbd/measure.rb:53]: ... lower-case version.

File: lib/measures/tbd/resources/psi.rb [hunk 3]
```diff
@@ -147,16 +138,18 @@ def append(p)
...
+    answer = false unless @set[s].has_key?(:rimjoist)
+    answer = false unless @set[s].has_key?(:parapet)
+    answer = false unless @set[s].has_key?(:fenestration)
+    answer = false unless @set[s].has_key?(:concave)
+    answer = false unless @set[s].has_key?(:convex)
+    answer = false unless @set[s].has_key?(:balcony)
+    answer = false unless @set[s].has_key?(:party)
+    answer = false unless @set[s].has_key?(:grade)
+    answer
   end
```
Comment (rd2) [inline lib/measures/tbd/resources/psi.rb:153]: Return boolean for ? method.

File: lib/measures/tbd/resources/psi.rb [hunk 4]
```diff
@@ -284,45 +277,43 @@ def processTBDinputs(surfaces, edges, set, io_path = nil, schema_path = nil)
...
       end
     end
-
   else
     # No (optional) user-defined TBD JSON input file.
     # In such cases, "set" must refer to a valid PSI set
-    psi.complete?(set)
+    raise "Incomplete PSI set #{set}" unless psi.complete?(set)
     io[:unit] = [{ psi: set }]       # i.e. default PSI set & no KHI's
   end
```
Comment (rd2) [inline lib/measures/tbd/resources/psi.rb:319]: The 40-odd line block needed a redesign to accommodate multiple edge matches.

File: lib/measures/tbd/resources/psi.rb [hunk 5]
```diff
@@ -924,17 +915,19 @@ def processTBD(os_model, psi_set, io_path = nil, schema_path = nil)
...
-    p = io[:unit].first[:psi]                                     # default PSI
-    p = edge[:io_set]         if match && edge.has_key?(:io_set)  # custom PSI
-    psi[t] = io_p.set[p][t]   if match && io_p.set[p][t]
+    if edge.has_key?(:io_type) # customized edge in TBD JSON file
+      match = true
+      t = edge[:io_type]
+      p = edge[:io_set]       if edge.has_key?(:io_set)
+      edge[:set] = p          if io_p.set.has_key?(p)
+      psi[t] = io_p.set[p][t] if io_p.set[p].has_key?(t)
+    end
```
Comment (rd2) [inline lib/measures/tbd/resources/psi.rb:928]: No actual change in the end, but this took many forms during the debugging process.

File: lib/measures/tbd/resources/psi.rb [hunk 6]
```diff
@@ -1052,6 +1045,7 @@ def processTBD(os_model, psi_set, io_path = nil, schema_path = nil)
     end # edge's surfaces loop
 
     edge[:psi] = psi unless psi.empty?
+    edge[:set] = p unless psi.empty?
```
Comment (rd2) [inline lib/measures/tbd/resources/psi.rb:1048]: A remedy - PSI sets were written out as PSI values in the output file.

File: lib/measures/tbd/resources/psi.rb [hunk 7]
```diff
@@ -1102,6 +1096,7 @@ def processTBD(os_model, psi_set, io_path = nil, schema_path = nil)
                 edge[:psi][t] = io_p.set[p][t] if io_p.set[p].has_key?(t)
               end
             end
+            edge[:set] = p
           end
```
Comment (rd2) [inline lib/measures/tbd/resources/psi.rb:1100]: See previous comment.

File: lib/measures/tbd/resources/psi.rb [hunk 8]
```diff
@@ -1391,13 +1387,43 @@ def processTBD(os_model, psi_set, io_path = nil, schema_path = nil)
...
+    next unless e.has_key?(:set)
+    v = e[:psi].values.max
+    p = e[:set]
+    next unless v > 0.000
+    t = e[:psi].key(v)
     l = e[:length]
-    edge = { psi: p, type: t, length: l, surfaces: e[:surfaces].keys }
-    io[:edges] << edge
+    edge = { psi: p, type: t, length: l, surfaces: e[:surfaces].keys, count: 1 }
+
```
Comment (rd2) [inline lib/measures/tbd/resources/psi.rb:1397]: A count (e.g. 1, 2, 3, ...) is indeed required for multiple edges (e.g. 4 sides of a skylight)

File: spec/tbd_tests_spec.rb [hunk 9]
```diff
@@ -1582,6 +1582,177 @@
...
 end
 
+RSpec.describe TBD do
+  it "can process TB & D : DOE Prototype test_warehouse.osm + JSON I/O" do
+    translator = OpenStudio::OSVersion::VersionTranslator.new
+    path = OpenStudio::Path.new(File.dirname(__FILE__) + "/files/test_warehouse.osm")
+    os_model = translator.loadModel(path)
+    expect(os_model.empty?).to be(false)
+    os_model = os_model.get
+
```
Comment (rd2) [inline spec/tbd_tests_spec.rb:1592]: A TBD JSON output-as-input test. In a nutshell, if everything is spot on, the output file of a first run can be reused as the input file of a subsequent run (both output files should be identical).

File: spec/tbd_tests_spec.rb [hunk 10]
```diff
@@ -1582,6 +1582,177 @@
...
+    out2 = JSON.pretty_generate(io2)
+    out_path2 = File.dirname(__FILE__) + "/../json/tbd_warehouse2.out.json"
+    File.open(out_path2, "w") do |out_path2|
+      out_path2.puts out2
+    end
+
+    # Both output files should be the same ...
+    cmd = "diff #{out_path} #{out_path2}"
+    expect(system( cmd )).to be(true)
+  end
```
Comment (rd2) [inline spec/tbd_tests_spec.rb:1753]: @macumber : Is it considered legit to run system commands as tests?

File: tbd.schema.json [hunk 11]
```diff
@@ -180,6 +180,12 @@
           },
           "minItems": 1,
           "uniqueItems": true
+        },
+        "count": {
+          "title": "Number of occurrences",
+          "type": "number",
+          "minimum": 0,
+          "exclusiveMinimum": true
         }
```
Comment (rd2) [inline tbd.schema.json:189]: Taking into consideration the new _edge_ counter.

File: lib/measures/tbd/resources/psi.rb [hunk 12]
```diff
@@ -147,16 +138,18 @@ def append(p)
...
-    raise "'#{s}' parapet?"         unless @set[s].has_key?(:parapet)
-    raise "'#{s}' fenestration?"    unless @set[s].has_key?(:fenestration)
-    raise "'#{s}' concave?"         unless @set[s].has_key?(:concave)
-    raise "'#{s}' convex?"          unless @set[s].has_key?(:convex)
-    raise "'#{s}' balcony?"         unless @set[s].has_key?(:balcony)
-    raise "'#{s}' party?"           unless @set[s].has_key?(:party)
-    raise "'#{s}' grade?"           unless @set[s].has_key?(:grade)
+  def complete?(s) # true/false
+    answer = true
+    answer = false unless @set.has_key?(s)
```
Comment (macumber) [inline lib/measures/tbd/resources/psi.rb:143]: I would make each of these:

`answer = (answer && @set.has_key?(s))`

Right now answer could get set to false (missing rimjoist) but then set back to true (has party)

File: spec/tbd_tests_spec.rb (hunk 9 above)
Comment (macumber) [reply spec/tbd_tests_spec.rb:1592]: Nice!

File: spec/tbd_tests_spec.rb (hunk 10 above)
Comment (macumber) [reply spec/tbd_tests_spec.rb:1753]: Yes, only issue here is that `diff` might not be present on some Windows machine (I have it because I have Git Bash)

Comment (macumber) [reply spec/tbd_tests_spec.rb:1753]: Pure Ruby way could be to read both files in as strings and do a string equality check

Comment (rd2) [reply spec/tbd_tests_spec.rb:1753]: OK. Good points. Will look into this (follow up commit).

**Pull request merged**